### ISSUE TITLE
UI freezes when showing large numbers of annotations in Java files

### ISSUE THREAD

**Issue by HeikoKlare** [eclipse-jdt/eclipse.jdt.ui#665]
## Description
The Java editor uses the `AnnotationModel` to show different types of annotations, such as the occurrences of the currently selected element. In case a file is quite large, this calculation can take quite long, for example, when changing the text selection to a different method.

## Reproduction
As an example, you can open the Java file in the attachment below and change the text selection between the `oftenCalledMethod` and other elements. This is an artificial and really extreme case, just for demonstration purposes. In our actual project, we experience the same problem (with even more performance problems) in large generated files with complex type and call hierarchies. 
[Test.zip](https://github.com/eclipse-jdt/eclipse.jdt.ui/files/12004611/Test.zip)

## Expected Behavior
The Java code editor should still be responsive when showing large source code files. In particular, navigation through the file should still happen in a responsive way.

## Reasons
The reason for the behavior is an inefficient implementation of the replacement of annotations in combination with the fact that the update of the annotation model requires a lock that blocks the UI, even though it is executed in a background job.
The replacement of annotations in the `AnnotationModel` (as done when changing the selection in an editor) processes every single annotation to be removed and added on its own. The Java-specific `CompilationUnitAnnotationModel` in the `CompilationUnitDocumentProvider` uses a `ReverseMap` that relies on an inefficient lookup within an `ArrayList` for every annotation to be added, removed or retrieved. The effort grows super-linear with the number of annotations to process.

## Possible Improvement
When performing a replacement of annotations, they should not be treated as individual removals and additions but as a batch operation. Then a temporary hash-based lookup structure can be created during the annotation model update to avoid the inefficient lookups in the currently used `ArrayList`.

I will propose a fix that implements such an improved annotation replacement.

**Comment by iloveeclipse** [comment]
> I will propose a fix that implements such an improved annotation replacement.

- Are we talking about `org.eclipse.jface.text.source.AnnotationModel` ? If yes, shouldn't the fix / the issue be reported in platform?
- Is this same as https://github.com/eclipse-platform/eclipse.platform.ui/issues/892 ? See discussion on **reverted** [fix for that](https://github.com/eclipse-platform/eclipse.platform.text/pull/22).

**Comment by HeikoKlare** [comment]
It is related, but not the same. JDT uses a subclass of `org.eclipse.jface.text.source.AnnotationModel` and has a custom implementation called `ReverseMap` for assigning values (in this case annotations) to positions in text (i.e., the reverse of the annotation map within the annotation model) . It implements a rather inefficient lookup of annotations since positions can change over time while editing the file.
The reason that this perform so bad is rooted in the platform implementation of `AnnotationModel.replaceAnnotations`, which is realized by processing every single addition and removal of annotations to replace. I also tried to improve that behavior with a batch replacement, but it is so intrusive that I gave up on that path and worked on a downstream patch.
In addition, the platform behavior is totally valid and the reason for the issue reported here is that the JDT-specific implementation performs so bad due to the custom map that is so inefficient in handling growing numbers of annotations.

Still, thanks for pointing to the fix you tried for the platform UI! It is at least related and may give some valuable insights for this issue.
